### DISCUSSION TITLE
Fehler beim Bauen FW für 7412 - make schlägt fehl ("Fehler" gefunden) und keine AVM-GUI ("Fehler" ebenfalls gefunden)

### DISCUSSION THREAD

**Discussion by MaxineMuster** [Freetz-NG/freetz-ng#864] [Allgemein] [answered]
Hallo zusammen,

nach langer Zeit musste ich für meine 7412 mal wieder ein Freetz-Image bauen, nachdem die Version 06.88 rauskam.

Das führte leider zu zwei Problemen (bereits "nacktem" Freetz, nur die 7412 gewählt, nichts weiteres) :

Ein frischer clone scheitert beim Bau an doppelten Definitionen:

```
In file included from ./include/istream:23:0,
                 from ./include/complex:20,
                 from src/complex.cpp:19:
./include/istream_helpers:357:44: error: redefinition of 'class std::__istream_readin<traits, char, long long int>'
  template <class traits> class _UCXXEXPORT __istream_readin<traits, char, long long>{
                                            ^
./include/istream_helpers:304:44: error: previous definition of 'class std::__istream_readin<traits, char, long long int>'
  template <class traits> class _UCXXEXPORT __istream_readin<traits, char, long long int>{
                                            ^
./include/istream_helpers:384:44: error: redefinition of 'class std::__istream_readin<traits, char, long long unsigned int>'
  template <class traits> class _UCXXEXPORT __istream_readin<traits, char, unsigned long long>{
                                            ^
./include/istream_helpers:329:44: error: previous definition of 'class std::__istream_readin<traits, char, long long unsigned int>'
  template <class traits> class _UCXXEXPORT __istream_readin<traits, char, unsigned long long int>{
                                            ^
make[2]: *** [Rules.mak:582: src/complex.o] Fehler 1
make[2]: *** Auf noch nicht beendete Prozesse wird gewartet …
```

Das läßt sich umgehen, indem man den Patch, der diese Definitionen reinbringt, rausnimmt:

```
mv make/libs/uclibcxx/patches/130-istream_long_long_support.freetz.uclibcxx.patch make/libs/uclibcxx/patches/130-istream_long_long_support.freetz.uclibcxx.patch.disable

 rm -rf source/target-mips_gcc-4.8.5_uClibc-0.9.33.2-nptl_kernel-3.10/uclibcxx-6687fc9276fa52defaf8592f2001c19b826aec93
 
```
 dann läuft es durch.
 
 
Zweites großes Problem: Mit dem gebauten Image gibt es keine AVM-GUI.

Ursache: ctlmgr (der Wrapper) startet nicht.

```
root@fritz:/var/mod/root# ctlmgr
/usr/bin/avm/ctlmgr: can't load library 'libctlmgr.so'
root@fritz:/var/mod/root# 
root@fritz:/var/mod/root# mount --bind /usr/bin/avm/ctlmgr /usr/bin/ctlmgr
root@fritz:/var/mod/root# ctlmgr
root@fritz:/var/mod/root# ps | grep ctl[m]
11787 root     18932 S    ctlmgr
root@fritz:/var/mod/root# 
```

Es wird für den ctlmgr ein Wrapper mit preload von "libctlmgr.so" eingebaut. Nur: libctlmgr.so gibt es nicht.
Also darf patches/scripts/105-wrapper_ctlmgr.sh hier nicht angewandt werden.

Ich habe das so "korrigiert", k.A. ob das so richtig war:

```
diff --git a/config/avm/features.in b/config/avm/features.in
index ddec729c3..4fd6fb1dd 100644
--- a/config/avm/features.in
+++ b/config/avm/features.in
@@ -293,7 +293,7 @@ config FREETZ_AVM_HAS_AVMMULTID_PRELOAD
 config FREETZ_AVM_HAS_AVMCTLMGR_PRELOAD
        bool
        default y
-       depends on !FREETZ_AVM_VERSION_07_2X
+       depends on !FREETZ_AVM_VERSION_07_2X && !(FREETZ_TYPE_7412 && FREETZ_AVM_VERSION_06_8X)
        help
                This is not mandatory for FOS 7.2X
 
```

In der alten Freetz-Version auf Basis von 06.87 ist zwar ebenfalls kein libctlmgr.so, aber in meinem Freetz von Ende 2021 gab es diesen Wrapper wohl auch noch nicht.

Oder bezieht sich das auf FREETZ_LIB_libctlmgr ? Die wird nämlich erst ab FREETZ_AVM_VERSION_07_0X_MIN geladen.
Aber bei FREETZ_AVM_HAS_AVMMULTID_PRELOAD ein "depends on" für FREETZ_LIB_libctlmgr ergäbe einen Zirkelschluss.
Muss vielleiche dort einfach auch ein "FREETZ_AVM_VERSION_07_0X_MIN" hin?

Vielleicht kann/müsste man das auch noch bei anderen Firmwares prüfen?

Danke für euren Einsatz!

**Comment by pfichtner** [comment]
Der uclibcxx Fehler tritt seit dem 17.09. (Upgrade von uclibcxx) auf, allerdings nur mit gcc-4.x auf (was bei mir bei mehreren Boxen die einzig mögliche gcc Version darstellt), mit gcc-5.x baut es (dort, wo gcc-5.x auswählbar ist)

**Comment by MaxineMuster** [comment]
Ah, danke, dann biete ich mal diesen Patch an, der für mich beides behebt:


```
diff --git a/config/avm/features.in b/config/avm/features.in
index ddec729c3..d4b340522 100644
--- a/config/avm/features.in
+++ b/config/avm/features.in
@@ -293,7 +293,7 @@ config FREETZ_AVM_HAS_AVMMULTID_PRELOAD
 config FREETZ_AVM_HAS_AVMCTLMGR_PRELOAD
 	bool
 	default y
-	depends on !FREETZ_AVM_VERSION_07_2X
+	depends on !FREETZ_AVM_VERSION_07_2X && ((FREETZ_AVM_VERSION_07_0X_MIN) || (FREETZ_AVM_VERSION_07_5X_MIN || FREETZ_AVM_HAS_USB_HOST))
 	help
 		This is not mandatory for FOS 7.2X
 
diff --git a/make/libs/uclibcxx/patches/130-istream_long_long_support.freetz.uclibcxx.patch b/make/libs/uclibcxx/patches/130-istream_long_long_support.freetz.uclibcxx.patch
index c939fc95e..39aa04cf3 100644
--- a/make/libs/uclibcxx/patches/130-istream_long_long_support.freetz.uclibcxx.patch
+++ b/make/libs/uclibcxx/patches/130-istream_long_long_support.freetz.uclibcxx.patch
@@ -4,7 +4,7 @@
  #ifdef __UCLIBCXX_EXPAND_ISTREAM_CHAR__
  #ifndef __UCLIBCXX_COMPILE_ISTREAM__
  
-+#ifndef __STRICT_ANSI__
++#if !defined __STRICT_ANSI__ && __C__ > 4
 +
 +//Support for input of long long data types
 +
@@ -22,7 +22,7 @@
 +	return is;
 +}
 +
-+#endif // __STRICT_ANSI__
++#endif // ! __STRICT_ANSI__ && __C__ > 4
  
  	template <> _UCXXEXPORT istream & basic_istream<char, char_traits<char> >::get(char & c);
  	template <> _UCXXEXPORT istream::int_type basic_istream<char, char_traits<char> >::get();
@@ -32,7 +32,7 @@
  		}
  	};
  
-+#ifndef __STRICT_ANSI__
++#if !defined __STRICT_ANSI__ && __C__ > 4
 +	template <class traits> class _UCXXEXPORT __istream_readin<traits, char, long long int>{
 +	public:
 +		inline static void readin(basic_istream<char, traits >& stream, long long int & var)
@@ -82,7 +82,7 @@
 +			}
 +		}
 +	};
-+#endif // __STRICT_ANSI__
++#endif // ! __STRICT_ANSI__ && __C__ > 4
 +
  
  #if !defined(__STRICT_ANSI__) || (__cplusplus >= 201103L)

```
Der erste wählt `FREETZ_AVM_HAS_AVMMULTID_PRELOAD` nur, wenn auch irgendeine der `FREETZ_LIB_libctlmgr*` Optionen gewählt wird.
Der zweite setzt den Patch außer Kraft, falls eine GCC 4 Version genutzt wird. Das kann man sicher noch granularer machen, aber dafür kenne ich das zu wenig...

**Comment by fda77** [comment]
Ich hätte den Patch der uclibcxx in ein Unterverzeichnis geschoben und in der .mk diesen conditional angewendet. Das macht es einfacher für komplexere Bedingungen oder Dinge die man beim compilieren nicht weiss. Aber so ists auch schön, danke :)
Mit den libctlmgr/libmultid Sachen bin ich noch nicht ganz fertig, da soll noch service-Support rein. Daher ist es da momentan noch etwas chaotisch...

* dfac4885b uclibcxx: fix gcc 4 and long-long, by @MaxineMuster
* 4911af166 fix FREETZ_AVM_HAS_AVMCTLMGR_PRELOAD by @MaxineMuster

**Reply by fda77** [reply]
Die lib sollte gar nicht mehr mit aktuellem fos nötig sein
* 4ce8de64d no libctlmgr for fos 7.5

**Comment by MaxineMuster** [comment, marked as answer]
Besten Dank fürs Fixen!

**Reply by fda77** [reply]
Es waren doch deine Patches!
